e2e/actions.go

```diff
@@ -0,0 +1,120 @@
+package e2e
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"time"
+)
+
+type Scenario struct {
+	actions []Action
+}
+
+func NewScenario() *Scenario {
+	return &Scenario{
+		actions: make([]Action, 0),
+	}
+}
+
+func (s *Scenario) AddAction(action Action) *Scenario {
+	s.actions = append(s.actions, action)
+	return s
+}
+
+func (s *Scenario) CleanUp(cluster *Cluster) {
+	for _, action := range s.actions {
+		action.Revert(cluster)
+	}
+}
+
+type Action interface {
+	Apply(c *Cluster)
+	Revert(c *Cluster)
+}
+
+// This action encapsulates logic for dropping nodes action.
+type DropNodeAction struct {
+	dropCount                int
+	dropProbabilityThreshold int
+	// TODO: Idea: Extract these two fields to another action (called PeriodicAction), which would receive as a parameter another action and would run it?
```

Comment:
I like the idea of modularity and smart design but lets make sure that going into this road of `Actions` inside `Actions` does not become a problem in the end. As far as I can think of, we only have three attack vectors? Drop node, latency and partition (flow map here). I can't think of anything else.

Reply:
The goal of creating `PeriodicAction` was just to simplify `DropNodeAction`, because `droppingInterval` and `duration` doesn't seem like they belong to the `DropNodeAction` (which is intended to be as simple as possible). Periodical execution of actions should be handled somehow from the outside (either by encapsulating that logic into the `PeriodicAction` struct or to let `Runner` take care of that etc.)